lib/hash_sockaddr.c

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/**
+ * @file hash_sockaddr.c
+ *
+ * @brief @c struct @c sockaddr hash related functions.
+ *
+ * Copyright (c) 2022, Intel Corporation
+ */
+
+#include <string.h>
+#include <netinet/in.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include <ell/util.h>
+#pragma GCC diagnostic pop
+
+
+#include <mptcpd/private/murmur_hash.h>
+
+#include "hash_sockaddr.h"
+
+
+/**
+ * @brief Bundle IPv4 address and port as a hash key.
+ */
+struct endpoint_in
+{
+        in_addr_t addr;
+        in_port_t port;
+};
+
+/**
+ * @brief Bundle IPv6 address and port as a hash key.
+ */
+struct endpoint_in6
+{
+        struct in6_addr addr;
+        in_port_t port;
+};
+
+/**
+ * @brief IPv4 address/port hash key buffer.
+ *
+ * This union ensures that padding bytes are initialized to zero in
+ * the @c endpoint member.  For this to work the @c buffer member must
+ * initialized prior to the @c endpoint member, e.g.
+ * @code
+ * union key_in key = { .buffer = { 0 } };
+ *
+ * key.endpoint.addr = sa->sin_addr.s_addr;
+ * key.endpoint.port = sa->sin_port;
+ * @endcode
+ * The goal is to avoid hashing uninitialized bytes in the hash key.
+ */
+union key_in
+{
+        struct endpoint_in endpoint;
+        uint8_t buffer[sizeof(struct endpoint_in)];
+};
+
+/**
+ * @brief IPv6 address/port hash key buffer.
+ *
+ * This union ensures that padding bytes are initialized to zero in
+ * the @c endpoint member.  For this to work the @c buffer member must
+ * initialized prior to the @c endpoint member, e.g.
+ * @code
+ * union key_in6 key = { .buffer = { 0 } };
+ *
+ * memcpy(key.endpoint.addr.s6_addr,
+ *        sa->sin6_addr.s6_addr,
+ *        sizeof(key.endpoint.addr.s6_addr));
+ *
+ * key.endpoint.port = sa->sin6_port;
+ * @endcode
+ * The goal is to avoid hashing uninitialized bytes in the hash key.
+ */
+union key_in6
+{
+        struct endpoint_in6 endpoint;
+        uint8_t buffer[sizeof(struct endpoint_in6)];
+};
+
+
+static unsigned int mptcpd_hash_sockaddr_in(struct sockaddr_in const *sa,
+                        uint32_t seed)
+{
+        // No port set.  Don't bother including it in the hash.
+        if (sa->sin_port == 0)
+                return mptcpd_murmur_hash3(&sa->sin_addr.s_addr,
+                                           sizeof(sa->sin_addr.s_addr),
+                                           seed);
```

Comment:
It's probably not worth including this zero-port case, and is probably a premature optimization.

Reply:
Agreed - I think it's simpler to avoid the extra code path and include the `0` port in the hash.

Reply:
Done.